### DISCUSSION TITLE
Estimated Memory Needed

### DISCUSSION THREAD

**Discussion by ghost** [idaholab/moose#27048] [Q&A General]
Is there a way within a MOOSE app to do some memory estimation ahead of the simulation? In ansys runs it will calculate the memory needed for a solve to fit entirely in memory or if it needs to switch to writing to disk for the simulation.

What I have run into is simulations that will start in a configuration on the HPC and exit with code 9, which I believe is an out of memory error. Sometimes when I adjust the configuration to use less cpus per node and more nodes to get more memory the solve will run. Other times it will still not solve and still exit with an error code 9.

It would be nice if I could get an estimate during the MOOSE setup where it determines how much memory it thinks it will need so that I can determine if the solve was never going to work or if there was potentially a memory leak somewhere.

**Comment by GiudGiud** [comment]
Hello

We do not have that implemented. You can get a view of the current memory expenditure from the perf graph live print. 
Code 9 is indeed OOM. 
Reducing the number of processes, using distributed meshes, removing auxiliary fields that are not needed are all good ways to reduce the memory footprint rapidly. For further improvements, you'll want to work on the preconditioning cost most of the time. 

> or if there was potentially a memory leak somewhere.

if you have a cluster with valgrind installed, it should be fairly easy to detect most memory leaks with that. 

We should write this down as a feature request. I think we can get a ball park estimate fairly easily, but it wont be very accurate without a lot of work interfacing with petsc to predict its needs

Guillaume

**Reply by ghost** [reply]
Ok thanks...is there a place to do the feature request?

**Reply by YaqiWang** [reply]
You can try `Debug/show_actions=true`, which will print memory usage at the beginning of each action (or setup stage). Particular attention should be given on `init_problem` task, where memory for solution vectors is allocated. If your `Executioner/solve_type` is a type other than JFNK, memory for preconditioning matrix will be allocated at its first evaluation (after MOOSE prints residual norm of the first linear iteration I believe). You can estimate memory usage of the matrix based on the number of unknowns of the nonlinear system printed by MOOSE. There could be other places of heavy memory usage depending on your application.

**Reply by GiudGiud** [reply]
Github issues are the place for feature requests.
